### PR TITLE
IOperation API review

### PR DESCRIPTION
**Pull request by JohnHamby**
This change set includes IOperation API changes suggested by a formal API review, as well as thoughtful independent reviews by @AnthonyDGreen, @nguerrera, @genlu, Don Syme, and probably others. There are no deliberate changes in behavior beyond what is necessary to conform to the new API, but there is some additional testing.

This is a rather large change set. The changes are almost entirely mechanical, but there are a lot of them.

Summary of the changes:
1)	Introduced SimpleUnaryOperationKind and SimpleBinaryOperationKind to describe unary and binary operations from an operand-type-independent perspective. Introduced UnaryOperandKind and BinaryOperandsKind to describe unary and binary operations from an operation-independent perspective. Introduced UnaryAndBinaryOperationsExtensions class to provide convenient access to these. Replaced occurrences of “Operator” with “OperatorMethod” in enumerations.
2)	Moved IOperation and OperationKind from Microsoft.CodeAnalysis.Semantics to Microsoft.CodeAnalysis.
3)	Removed IExpressionExtensions.
4)	Renamed IExpression.ResultType to “Type”. I’m not crazy about this change.
5)	Renamed ILiteralExpression.Spelling to “Text”.
6)	Renamed ILockStatement.Locked to “LockedObject”.
7)	Introduced IHasArgumensExpression as a common base interface of IInvocationExpression and IObjectCreationExpression.
8)	Renamed IHasArgumentsExpression.ArgumentMatchingParameter to “GetArgumentMatchingParameter”.
9)	Renamed IThowStatement.Thrown to”ThrownObject”.
10)	Renamed IUsingWithDeclaration.Variables to “Declaration”.
11)	Removed ITypeOperationExpression.TypeOperationKind. Introduced ISizeOfExpression and ITypeOfExpression.
12)	Renamed IVariable to “IVariableDeclaration”.
13)	Added Compilation property to OperationBlockAnalysisContext and OperationBlockStartAnalysisContext.
14)	Renamed IReturnStatement.Returned to “ReturnedValue”.
15)	Removed IStatement.
16)	Replaced operation kinds GoToStatement, BreakStatement, and ContinueStatement with BranchStatement. Added BranchKind to IBranchStatement.
17)	Renamed ConversionKind.AsCast to “TryCast”.
18)	Renamed ICatch to “ICatchClause”. Renamed OperationKind.CatchHandler to “CatchClause”.
19)	Renamed ICase to “ISwitchCase”. Renamed OperationKind.SwitchSection to “SwitchCase”.
20)	Removed ILabeledStatement and OperationKind.LabeledStatement.
21)	Reworked IInstanceReferenceExpression to no longer expose the this parameter symbol.
22)	Removed IExpression. Moved members of IExpression to IOperation.

@genlu @nguerrera @mavasani @srivatsn @ManishJayaswal @mattwar @jaredpar @terrajobst @AnthonyDGreen participated in the review and might glance at the outcome.

### CODE REVIEW

**Comment by JohnHamby** [comment]
@dotnet-bot test this please.

**Comment by JohnHamby** [comment]
@dotnet-bot retest this.

**Comment by nguerrera** [comment]
> 4) Renamed IExpression.ResultType to “Type”. I’m not crazy about this change.

Now that it's ultimately IOperation.Type, I think I agree that ResultType is better. To me, it was natural for 'result' to be implicit on an expression, but I don't feel the same way now that it's associated with any operation.

**Comment by nguerrera** [comment]
I have some concerns about the IExpression and IStatement -> IOperation merger:

For IL -> IOperation, there two things that I imagined would be helped by this:

1. Removing the need to wrap assignments and invocations into ExpressionStatements.
2. Opening a window for associating an arbitrary block with an exception filter

Re (1): There are still ExpressionStatement nodes here, so it's unclear if it's now OK to just put bare assignment and invocation expressions in to a statement list.

Re (2): I can now put an IBlockStatement in an ICatchClause.Filter, but still need to convey where the boolean filter result is obtained from the block. I could imagine saying that the result of the final operation in a block is the result of the block, but going down that route would suggest to me that we should be pushing the envelope further on the unification of expressions and statements...

For example, there are now two structurally equivalent conditional nodes: IConditionalChoiceExpression and IIfStatement. In a semantic model where everything can have a result (and imagining representing languages where that is the case), I would not expect this dichotomy.

I think that if there are expression statements and distinct expression/statement analogs (I can imagine more than just if/then/else vs. ternary over time as more language features or languages are onboarded: switch expressions, try expressions, block expressions), then we should keep the strong-typing of IExpression/IStatement.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 1]
```diff
@@ -888,11 +896,9 @@ public override void Accept(OperationVisitor visitor)
 
     internal partial class BoundBaseReference : IInstanceReferenceExpression
     {
-        bool IInstanceReferenceExpression.IsExplicit => this.Syntax.Kind() == SyntaxKind.BaseExpression;
-
-        IParameterSymbol IParameterReferenceExpression.Parameter => (IParameterSymbol)this.ExpressionSymbol;
+        InstanceReferenceKind IInstanceReferenceExpression.InstanceReferenceKind => InstanceReferenceKind.BaseClass;
```
Comment (nguerrera) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Do we need this distinction between this/base or is it implied from the bound target method and whether the invocation is virtual?

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs [hunk 2]
```diff
@@ -16,6 +16,10 @@ internal partial class BoundStatement : IStatement
 
         SyntaxNode IOperation.Syntax => this.Syntax;
 
+        ITypeSymbol IOperation.Type => null;
```
Comment (nguerrera) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: I would have expected void here.

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs [hunk 3]
```diff
@@ -634,6 +658,9 @@ public override void Accept(OperationVisitor visitor)
 
     internal partial class BoundLabelStatement : ILabelStatement
     {
+        // These represent synthesized labels, and do not have an attached statement.
+        IOperation ILabelStatement.LabeledStatement => null;
```
Comment (nguerrera) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: Should we put a synthesized empty statement here?

File: src/Compilers/Core/Portable/Compilation/Expression.cs [hunk 4]
```diff
@@ -212,26 +220,26 @@ public void Accept(OperationVisitor visitor)
...
         private sealed class CompoundAssignmentExpression : ICompoundAssignmentExpression
         {
-            public CompoundAssignmentExpression(IReferenceExpression target, IExpression value, BinaryOperationKind binaryKind, IMethodSymbol operatorMethod, SyntaxNode syntax)
+            public CompoundAssignmentExpression(IReferenceExpression target, IOperation value, BinaryOperationKind binaryKind, IMethodSymbol operatorMethod, SyntaxNode syntax)
             {
                 this.Target = target;
                 this.Value = value;
-                this.BinaryKind = binaryKind;
-                this.Operator = operatorMethod;
+                this.BinaryOperationKind = binaryKind;
```
Comment (nguerrera) [inline src/Compilers/Core/Portable/Compilation/Expression.cs]: I'd make the parameter name match the new property name: binaryKind -> binaryOperationKind.

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs [hunk 5]
```diff
@@ -137,11 +149,11 @@ public override void Accept(OperationVisitor visitor)
 
     internal partial class BoundIfStatement : IIfStatement
     {
-        IExpression IIfStatement.Condition => this.Condition;
+        IOperation IIfStatement.Condition => this.Condition;
 
-        IStatement IIfStatement.IfTrue => this.Consequence;
+        IOperation IIfStatement.IfTrue => this.Consequence;
```
Comment (nguerrera) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: With the push to make all properties nouns, should we rename IfTrue and IfFalse? (I actually like IfTrue and IfFalse as they are more obvious than Consequence and Alternative, but I also liked Thrown, Locked, and Upon so I'm not sure my taste carries any weight over the guideline we're now trying to follow,)

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 1 above)
Comment (nguerrera) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I ask because in IL there are more bindings possible than this/base/MyClass can imply and I'm not sure what to answer for such cases.

File: src/Compilers/Core/Portable/Compilation/IOperation.cs [hunk 6]
```diff
@@ -35,95 +44,164 @@ public interface IOperation
...
+        TypeParameterObjectCreationExpression = 0x2d,
+        /// <summary>Indicates an <see cref="IArrayCreationExpression"/>.</summary>
+        ArrayCreationExpression = 0x2e,
+        /// <summary>Indicates an <see cref="IDefaultValueExpression"/>.</summary>
+        DefaultValueExpression = 0x2f,
+        /// <summary>Indicates an <see cref="IInstanceReferenceExpression"/>.</summary>
+        InstanceReferenceExpression = 0x30,
+        /// <summary>Indicates an <see cref="IIsExpression"/>.</summary>
+        IsExpression = 0x33,
+        // TypeOperationExpression = 0x34,
```
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: Do you think it worth to change the value of `SizeOfExpression` (to 35) and shift the following values?

File: src/Compilers/Core/Portable/Compilation/IOperation.cs [hunk 7]
```diff
@@ -35,95 +44,164 @@ public interface IOperation
...
+        /// <summary>Indicates an <see cref="IUsingWithExpressionStatement"/>.</summary>
+        UsingWithExpressionStatement = 0x14,
+        /// <summary>Indicates an <see cref="IReturnStatement"/>.</summary>
+        YieldReturnStatement = 0x15,
+        /// <summary>Indicates an <see cref="IFixedStatement"/>.</summary>
+        FixedStatement = 0x16,
+        // LocalFunctionStatement = 0x17,
+
+        /// <summary>Indicates an <see cref="IExpressionStatement"/>.</summary>
+        ExpressionStatement = 0x18,
```
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: I probably missed the discussion during review, but I think it'd be nice to leave some room between different type of operations for possible future expansion (i.e. statement vs. expression vs. neither, etc.)

File: src/Compilers/Core/Portable/Compilation/IOperation.cs [hunk 8]
```diff
@@ -25,6 +24,16 @@ public interface IOperation
         /// </summary>
         SyntaxNode Syntax { get; }
 
+        /// <summary>
+        /// Result type of the operation. Null if the operation is not a value-producing expression.
```
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: Is it just me or this sounds like sayings all operations are expression? Maybe change to soemthing like `Null if it is not a value-producing operation` or `Null if the operation is not an expression`

File: src/Compilers/Core/Portable/Compilation/IStatement.cs [hunk 9]
```diff
@@ -101,19 +94,19 @@ public enum CaseKind
         /// <summary>
         /// Indicates case x in C# or Case x in VB.
         /// </summary>
-        SingleValue,
+        SingleValue = 0x0,
```
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IStatement.cs]: FxCop suggests all the enums should have a None = 0x0 to indicate uninitialized. Should we enforce that?
https://msdn.microsoft.com/en-us/library/ms182149.aspx

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 10]
```diff
@@ -31,29 +32,17 @@ public interface IInvocationExpression : IExpression
         /// <summary>
         /// 'This' or 'Me' argument to be supplied to the method.
         /// </summary>
-        IExpression Instance { get; }
+        IOperation Instance { get; }
```
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Mentioned this with @JohnHamby offline, this would return null for calls to static method. Maybe worth mentioning in the comment.

Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Some people might consider `This` and `Me` as receiver instead of argument, the comment here would be confusing for them maybe?

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 11]
```diff
@@ -4,25 +4,26 @@
...
+        /// <param name="parameter">Parameter of the target method.</param>
+        /// <returns>Argument corresponding to the parameter.</returns>
+        IArgument GetArgumentMatchingParameter(IParameterSymbol parameter);
     }
 
     /// <summary>
     /// Represents a C# or VB method invocation.
     /// </summary>
-    public interface IInvocationExpression : IExpression
+    public interface IInvocationExpression : IHasArgumentsExpression
```
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: As of now, I think raising event is also covered by this. Right? If we don't separate it now, introducing a `RaiseEvent` operation might become a breaking change in the future.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 12]
```diff
@@ -298,18 +300,18 @@ public interface IEventAssignmentExpression : IExpression
...
     /// Represents a conditional access expression.
     /// </summary>
-    public interface IConditionalAccessExpression : IExpression
+    public interface IConditionalAccessExpression : IOperation
     {
         /// <summary>
         /// Expression subject to conditional access.
         /// </summary>
-        IExpression Access { get; }
+        IOperation Access { get; }
```
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: @JohnHamby Are you gonna address the problem mentioned in #8476 in another PR?

Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Yes.

**Comment by JohnHamby** [comment]
Re: @nguerrera 's comments on the elimination of IExpression and IStatement:

There are now no baked-in structural constraints about where an expression can appear or where a statement can appear. It is structurally OK for expressions to occur in a block's list of operations. It is structurally OK for a non-value-producing operation to occur as the operation of a binary operation. Doing those things might be a natural expression for some languages. However, the structural constraints of VB and C# indicate a different representation--the existence of IExpressionStatement and the dichotomy between if statements and conditional expressions is appropriate for these languages. It's possible to blur the distinctions, and most analyzers would (by design) never notice, but my inclination is to preserve the structure of source languages where possible.

Concretely, if we preserve an expression/statement distinction, we can't represent F#. Having the ability to represent an everything-is-an-expression language does not need to influence how we represent VB and C#.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: @nguerrera is correct that the this/base/class distinction can be inferred. Is the inference too subtle for analyzer authors?

What are the other possibilities in IL? Can they be generated from C# or VB?

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs (hunk 3 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: No. This was as close as I could get to unifying the treatment of labels in C# and VB. This property is mostly to make the walker work.

For the C# fragment:
    Target: if (...);
    return 3;
the C# compiler generates 2 statements in the containing block. The first is a labeled statement that has an if statement as an operand. The second is a return statement.

For the equivalent VB fragment, the VB compiler generates 3 statements in the containing block. The first is a labeled statement with no operand, the second is an if statement, and the third is a return statement.

An analyzer should generally ignore the statement operand of a label statement, but the walker needs access to it in order to expose the labeled statement to analyzers.

File: src/Compilers/Core/Portable/Compilation/IOperation.cs (hunk 7 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: Yes, I agree. I ran out of brain before attempting that, and we can do it in a considered fashion in a follow-up PR.

File: src/Compilers/Core/Portable/Compilation/IOperation.cs (hunk 6 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: No, given that we should probably reorganize the values entirely.

File: src/Compilers/Core/Portable/Compilation/IStatement.cs (hunk 9 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IStatement.cs]: Yes. I'll make that change.

File: src/Compilers/Core/Portable/Compilation/Expression.cs (hunk 4 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/Expression.cs]: OK.

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs (hunk 5 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: You had to notice. How about "IfTrueValue" and "IfFalseValue"?

I liked the older names better too.

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs (hunk 2 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: I'm not sure that the predefined types are available from within the implementation of bound trees. I'm also prepared to defend the distinction between "could have a type that turns out to be void" and "cannot have a type". A null type implies the latter.

File: src/Compilers/Core/Portable/Compilation/IOperation.cs (hunk 8 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: How about "Null if the operation does not produce a result."?

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 10 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: How about "'This' or 'Me' instance to be supplied to the method, or null if the method is static."?

File: src/Compilers/Core/Portable/Compilation/IOperation.cs (hunk 8 above)
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: Sounds good.

File: src/Compilers/Core/Portable/Compilation/IStatement.cs (hunk 9 above)
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IStatement.cs]: Thanks!

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 10 above)
Comment (genlu) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Sounds good!

**Comment by JohnHamby** [comment]
Test prtest/win/dbg/unit32 please.

**Comment by JohnHamby** [comment]
I'm going to preemptively merge because other work is blocked, but will continue to monitor feedback and will make any corrections in a future PR.

**Comment by nguerrera** [comment]
> Concretely, if we preserve an expression/statement distinction, we can't represent F#. Having the ability to represent an everything-is-an-expression language does not need to influence how we represent VB and C#.

Would a tree for an everything-is-an-expression language be expected to reuse the XxxStatement nodes at all? e.g. Is a BlockStatement with non-null ResultType valid? Or would we add nodes like BlockExpression as such languages are onboarded? If the statement nodes can be reused, then the semantics would need to be documented: e.g. "When ResultType is non-null, evaluates to the result of the final IOperation in the block."

**Comment by AnthonyDGreen** [comment]
Wouldn't it be fair to say that "statements" are simply void-returning operations? The languages are already prescriptive about where void-type values can be - typically at the statement level. In VB I can't call a Sub in the middle of an argument list - it has to appear at the block level.

This is similar, though different, from F#'s unit type and could easily support that representation as well. 

Some more cases:
* Assignment is an expression in C# but a statement in VB. However, after initial binding in VB we also represent assignment as an expression because it makes describing some lowerings much easier.
* We considered adding "sequence expressions" to C# 6, and actually have them in the lowered trees. It may or may not show up in a future version of the language.
* In C# 7 we're deliberating making what is now a throw statement into a throw expression.
* Bound conditional expressions and the proposed C# 7 match expression could be said to have control flow. And a bound sequence expression can have side-effects that would otherwise be represented as the children of a block statement, so this blurring goes both ways.

In short, the distinction between statements and expressions is capricious. I think moving this distinction to the type of the operation rather than a structural constraint gives us far more wiggle room both for supporting new languages _and_ for evolving our current languages in the future.

P.S. Statements can't have a null return type. Typeless expressions like the Nothing literal or method groups are typeless. Statements that don't return a value would be better be represented as void-returning (as we do with invocation expressions that invoke Subs today).

**Comment by nguerrera** [comment]
> P.S. Statements can't have a null return type. 
This was my intuition as well, but see @JohnHamby's previous reply: https://github.com/dotnet/roslyn/pull/8624#discussion_r52812563

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 13]
```diff
@@ -181,29 +172,43 @@ public interface ISyntheticLocalReferenceExpression : IReferenceExpression
...
-        ForLoopLimitValue
+        ForLoopLimitValue = 0x2
     }
 
     /// <summary>
-    /// Represents a reference to a C# this or VB Me parameter.
+    /// Represents a C# this or base expression, or a VB Me, MyClass, or MyBase expression.
     /// </summary>
-    public interface IInstanceReferenceExpression : IParameterReferenceExpression
+    public interface IInstanceReferenceExpression : IOperation
```
Comment (nguerrera) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: I believe this should inherit from IReferenceExpression and not just IOperation since 'this' is effectively a byref argument for struct methods and can be assigned to.
